sql/src/SqlTables.sk

```diff
@@ -357,6 +357,176 @@ fun getLatestIndexDescr(
+        }
+      };
+      table.writeArraySourceMany(
+        context,
+        table.unsafeGetFileIter().flatMap(clearData),
+      );
+    });
+
+    _ = SKDB.evalSelect(context, newCSelect, None());
+
+    dirDescr = SKDB.Compiler::getDirDescr(
+      (selectFile.select.core as P.SelectCoreQuery _).params,
+      newCSelect.from,
+      newCSelect.params,
+      newCSelect.id,
+      dirName,
+      true,
+    );
+    tableDir.writeArray(context, fileName, Array[dirDescr]);
+  };
+}
+
+fun alterTableAddColumn(
+  context: mutable SKStore.Context,
+  pos: Int,
+  name: P.Name,
+  newCol: P.ColumnDefinition,
+  options: Options,
+): void {
+  if (
+    name == SKDBUsersByName::name ||
+    name == SKDBUserPermissions::name ||
+    name == SKDBGroups::name ||
+    name == SKDBGroupPermissions::name
+  ) {
+    error(pos, "Can't add column to predefined table")
+  };
+  if (newCol.notNull is Some _ && newCol.default is None _) {
+    error(pos, "ADD COLUMN must either be nullable or provide a default value");
+  };
+  if (newCol.unique is Some _ || newCol.primary is Some _) {
+    error(pos, "ADD COLUMN must not impose a uniqueness or PK constraint")
+  };
+
+  tableDir = getTableDir(context);
+  sid = SKStore.SID::create(name.lower);
+
+  oldDescr = tableDir.maybeGet(context, sid) match {
+  | None() -> error(pos, "Table '" + name + "' does not exist")
+  | Some(descr) -> getLatestDirDescr(context, descr)
+  };
+  dirName = oldDescr.dirName;
+  dir = context.unsafeGetEagerDir(dirName);
+  oldNCols = oldDescr.schema.size();
+
+  newSchema = Array::fillBy(oldNCols + 1, i ->
+    if (i < oldNCols) oldDescr.schema[i] else newCol
+  );
+  newKinds = newSchema.mapWithIndex((idx, col) -> (idx, P.IASC(), col.ty));
+
+  convertToNewSchema: (
+    (SKStore.Key, mutable Iterator<SKStore.File>),
+  ) -> mutable Iterator<
+    (SKStore.Key, (SKStore.Path, SKStore.Path, Array<SKStore.File>)),
+  > = (kv) -> {
+    (key, rowIter) = kv;
+    rowIter.next() match {
+    | Some(oldRow @ RowValues _) ->
+      repeat = oldRow.repeat;
+      rowIter.each(r ->
+        r match {
+        | RowValues(_, rr) -> !repeat = repeat + rr
+        | _ -> invariant_violation("Invalid SKDB row")
+        }
+      );
+
+      newVals = RowValues(
+        Array::fillBy(oldNCols + 1, i ->
+          if (i < oldNCols) oldRow.getValue(i) else {
+            newCol.default.flatMap(def -> evaluateDefault(context, def.expr))
+          }
+        ),
+        repeat,
+      );
+      newKey = RowKey(newVals, newKinds);
+      newSource = SKStore.Path::create(dirName, newKey);
+
+      // Since schema/kinds information is encoded in SKDB.RowKeys, we have to both null
+      // out the old row(s) and write the new row. (instead of just overwriting the old key)
+      for (x in dir.unsafeGetAllDataIter(key)) {
+        (_tick, source, writer, _row) = x;
+        yield (key, (source, writer, Array[]))
+      };
+      yield (newKey, (newSource, newSource, Array[newVals]))
```

Comment:
I'm scared of yield. I've seen some weird compiler issues with it and had to remove it from somewhere once (I forget now). Just pointing this out in case you see any super strange behaviour.

Reply:
Oh, good to know - will keep in mind!  My thinking was that composing iterators would be more lightweight (both in complexity and in actual allocations etc.) but can revisit if we see weirdness